events/equeue/equeue.c

```diff
@@ -418,6 +418,7 @@ void equeue_dispatch(equeue_t *q, int ms) {
                     q->background.active = true;
                     equeue_mutex_unlock(&q->queuelock);
                 }
+                q->break_requested = false;
```

Comment:
I feel a bit nervous about actually writing to this variable outside the mutex - one step worse than reading it. In both cases it has no synchronisation protection, which could in principle cause problems with multiple CPUs. Some C11/C++11 atomic would sort it out, but we don't have that available.

However, I can't actually construct a plausible failure mode looking at it - all the mutexes and semaphores that do exist around it add quite a lot of ordering constraints. And I guess we assume only 1 thread ever dispatches the event queue - we're synchronising between multiple queuers and 1 dispatcher, not multiple dispatchers?

Reply:
I did originally move it to be within the mutex block in 413-419, but of course that led to failing tests because that mutex  is only entered if there's an update object attached.

How about if I expand the mutex lock to outside the update conditional?  It hurts the normal case, but doesn't impact the worst case.

As in...

```
        // check if we should stop dispatching soon
        if (ms >= 0) {
            deadline = equeue_tickdiff(timeout, tick);
            if (deadline <= 0) {
                equeue_mutex_lock(&q->queuelock);
                // update background timer if necessary
                if (q->background.update) {
                    if (q->background.update && q->queue) {
                        q->background.update(q->background.timer,
                                equeue_clampdiff(q->queue->target, tick));
                    }
                    q->background.active = true;
                }
                q->break_requested = false;
                equeue_mutex_unlock(&q->queuelock);
                return;
            }
        }
```

Reply:
The original implementation seemed fine to me. The case we're concerned about is a break from inside the event queue, which by definition is synchronous.

With multiple dispatchers all we're gauranteed is that at _least_ one thread will break, which is already protected by a mutex. This race condition can only happen after a thread has already decided to exit, but even with a critical section it's still a race since you don't know if the break will be consumed before or after the timeout.

I'd be fine with extending the mutex to be safe, except it does look like it has a negative performance impact (using `make prof`):
``` bash
beginning profiling...
baseline_prof: 23 cycles (+0%)
equeue_tick_prof: 43 cycles (+0%)
equeue_alloc_prof: 61 cycles (+0%)
equeue_post_prof: 227 cycles (+0%)
equeue_post_future_prof: 227 cycles (+0%)
equeue_dispatch_prof: 312 cycles (-34%)   <-- ouch, was 232 cycles
equeue_cancel_prof: 121 cycles (+0%)
equeue_alloc_many_prof: 64 cycles (+0%)
equeue_post_many_prof: 221 cycles (+2%)
equeue_post_future_many_prof: 220 cycles (+3%)
equeue_dispatch_many_prof: 6952 cycles (+0%)
equeue_cancel_many_prof: 122 cycles (+0%)
equeue_alloc_size_prof: 56 bytes (+0%)
equeue_alloc_many_size_prof: 64000 bytes (+0%)
equeue_alloc_fragmented_size_prof: 64000 bytes (+0%)
done!
```

Although, this is on Linux, where mutex acquisition isn't cheap.